beacon-chain/sync/validate_sync_committee_message.go

```diff
@@ -0,0 +1,310 @@
+//
+//  sync_committee_{subnet_id}
+
+// The sync_committee_{subnet_id} topics are used to propagate unaggregated sync committee messages
+// to the subnet subnet_id to be aggregated before being gossiped to the
+// global sync_committee_contribution_and_proof topic.
+
+// The following validations MUST pass before forwarding the sync_committee_message on the network:
+
+// [IGNORE] The message's slot is for the current slot (with a MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance),
+// i.e. sync_committee_message.slot == current_slot.
+// [REJECT] The subnet_id is valid for the given validator, i.e. subnet_id in
+// compute_subnets_for_sync_committee(state, sync_committee_message.validator_index).
+// Note this validation implies the validator is part of the broader current sync
+// committee along with the correct subcommittee.
+// [IGNORE] There has been no other valid sync committee message for the declared slot for the
+// validator referenced by sync_committee_message.validator_index (this requires maintaining
+// a cache of size SYNC_COMMITTEE_SIZE // SYNC_COMMITTEE_SUBNET_COUNT for each subnet that can be
+// flushed after each slot). Note this validation is per topic so that for a given slot, multiple
+// messages could be forwarded with the same validator_index as long as the subnet_ids are distinct.
+// [REJECT] The signature is valid for the message beacon_block_root for the validator referenced by validator_index.
+func (s *Service) validateSyncCommitteeMessage(
+	ctx context.Context, pid peer.ID, msg *pubsub.Message,
+) pubsub.ValidationResult {
+	ctx, span := trace.StartSpan(ctx, "sync.validateSyncCommitteeMessage")
+	defer span.End()
+
+	if pid == s.cfg.P2P.PeerID() {
+		return pubsub.ValidationAccept
+	}
+
+	// Basic validations before proceeding.
+	if result := withValidationPipeline(
+		ctx,
+		s.ignoreIfSyncing(),
+		rejectNilTopic(msg),
+	); result != pubsub.ValidationAccept {
+		return result
+	}
+
+	// Read the data from the pubsub message, and reject if there is an error.
+	m, err := s.readSyncCommitteeMessage(msg)
+	if err != nil {
+		log.WithError(err).Debug("Could not decode message")
+		traceutil.AnnotateError(span, err)
+		return pubsub.ValidationReject
+	}
+
+	// Validate sync message times before proceeding.
+	if result := withValidationPipeline(
+		ctx,
+		s.ignoreInvalidSyncMsgTime(m),
+	); result != pubsub.ValidationAccept {
+		return result
+	}
+
+	committeeIndices, err := s.cfg.Chain.HeadCurrentSyncCommitteeIndices(ctx, m.ValidatorIndex, m.Slot)
+	if err != nil {
+		traceutil.AnnotateError(span, err)
+		return pubsub.ValidationIgnore
+	}
+
+	// Validate the message's data according to the p2p specification.
+	if result := withValidationPipeline(
+		ctx,
+		ignoreEmptyCommittee(committeeIndices),
+		s.rejectIncorrectCommittee(committeeIndices, *msg.Topic),
+		s.ignoreHasSeenSyncMsg(m, committeeIndices),
+		s.rejectInvalidSignature(m),
+	); result != pubsub.ValidationAccept {
```

Comment:
I may be misunderstanding this pipeline pattern, but why are there 3 pipelines in one method?

Do these validations have to be executed in order or depend on the result of the previous step?

Reply:
Moving forward, will only use pipelines if they can be immediately sequential and will try to have only 1 pipeline in the method